# Практическая работа

# Задача

Один из способов повысить эффективность взаимодействия банка с клиентами — отправлять предложение о новой услуге не всем клиентам, а только некоторым, которые выбираются по принципу наибольшей склонности к отклику на это предложение.

Задача заключается в том, чтобы предложить алгоритм, который будет выдавать склонность клиента к положительному или отрицательному отклику на предложение банка. Предполагается, что, получив такие оценки для некоторого множества клиентов, банк обратится с предложением только к тем, от кого ожидается положительный отклик.


Для решения этой задачи загрузите файлы из базы в Postgres.
Эта БД хранит информацию о клиентах банка и их персональные данные, такие как пол, количество детей и другие.

Описание таблиц с данными представлено ниже.


**D_work**

Описание статусов относительно работы:
- ID — идентификатор социального статуса клиента относительно работы;
- COMMENT — расшифровка статуса.


**D_pens**

Описание статусов относительно пенсии:
- ID — идентификатор социального статуса;
- COMMENT — расшифровка статуса.


**D_clients**

Описание данных клиентов:
- ID — идентификатор записи;
- AGE	— возраст клиента;
- GENDER — пол клиента (1 — мужчина, 0 — женщина);
- EDUCATION — образование;
- MARITAL_STATUS — семейное положение;
- CHILD_TOTAL	— количество детей клиента;
- DEPENDANTS — количество иждивенцев клиента;
- SOCSTATUS_WORK_FL	— социальный статус клиента относительно работы (1 — работает, 0 — не работает);
- SOCSTATUS_PENS_FL	— социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
- REG_ADDRESS_PROVINCE — область регистрации клиента;
- FACT_ADDRESS_PROVINCE — область фактического пребывания клиента;
- POSTAL_ADDRESS_PROVINCE — почтовый адрес области;
- FL_PRESENCE_FL — наличие в собственности квартиры (1 — есть, 0 — нет);
- OWN_AUTO — количество автомобилей в собственности.


**D_agreement**

Таблица с зафиксированными откликами клиентов на предложения банка:
- AGREEMENT_RK — уникальный идентификатор объекта в выборке;
- ID_CLIENT — идентификатор клиента;
- TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было).
    
    
**D_job**

Описание информации о работе клиентов:
- GEN_INDUSTRY — отрасль работы клиента;
- GEN_TITLE — должность;
- JOB_DIR — направление деятельности внутри компании;
- WORK_TIME — время работы на текущем месте (в месяцах);
- ID_CLIENT — идентификатор клиента.


**D_salary**

Описание информации о заработной плате клиентов:
- ID_CLIENT — идентификатор клиента;
- FAMILY_INCOME — семейный доход (несколько категорий);
- PERSONAL_INCOME — личный доход клиента (в рублях).


**D_last_credit**

Информация о последнем займе клиента:
- ID_CLIENT — идентификатор клиента;
- CREDIT — сумма последнего кредита клиента (в рублях);
- TERM — срок кредита;
- FST_PAYMENT — первоначальный взнос (в рублях).


**D_loan**

Информация о кредитной истории клиента:
- ID_CLIENT — идентификатор клиента;
- ID_LOAN — идентификатор кредита.

**D_close_loan**

Информация о статусах кредита (ссуд):
- ID_LOAN — идентификатор кредита;
- CLOSED_FL — текущий статус кредита (1 — закрыт, 0 — не закрыт).

Ниже представлен минимальный список колонок, которые должны находиться в итоговом датасете после склейки и агрегации данных. По своему усмотрению вы можете добавить дополнительные к этим колонки.

    - AGREEMENT_RK — уникальный идентификатор объекта в выборке;
    - TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было);
    - AGE — возраст клиента;
    - SOCSTATUS_WORK_FL — социальный статус клиента относительно работы (1 — работает, 0 — не работает);
    - SOCSTATUS_PENS_FL — социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
    - GENDER — пол клиента (1 — мужчина, 0 — женщина);
    - CHILD_TOTAL — количество детей клиента;
    - DEPENDANTS — количество иждивенцев клиента;
    - PERSONAL_INCOME — личный доход клиента (в рублях);
    - LOAN_NUM_TOTAL — количество ссуд клиента;
    - LOAN_NUM_CLOSED — количество погашенных ссуд клиента.


Будьте внимательны при сборке датасета: это реальные банковские данные, в которых могут наблюдаться дубли, некорректно заполненные значения или значения, противоречащие друг другу. Для получения качественной модели необходимо предварительно очистить датасет от такой информации.

## Задание 1

Соберите всю информацию о клиентах в одну таблицу, где одна строчка соответствует полной информации об одном клиенте.

#### Так как была дана возможность работать с файлами напрямую, так и сделаем

In [1]:
import pandas as pd

In [2]:
#Загружаем данные
df1 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_clients.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_close_loan.csv')
df3 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_job.csv')
df4 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_last_credit.csv')
df5 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_loan.csv')
df6 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_pens.csv')
df7 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_salary.csv')
df8 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_target.csv')
df9 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_work.csv')

In [3]:
#Объединяем данные
merged_df = df3.merge(df4, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df5, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df7, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df8, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df2, on='ID_LOAN', how='outer')
df1.rename(columns={'ID': 'ID_CLIENT'}, inplace=True)
merged_df = merged_df.merge(df1, on='ID_CLIENT', how='outer')

#Удаляем полные дубликаты и делаем бинарные переменные
merged_df = merged_df.drop_duplicates()

# Создание столбца LOAN_NUM_TOTAL
merged_df['LOAN_NUM_TOTAL'] = merged_df.groupby('ID_CLIENT')['ID_LOAN'].transform('count')
# Создание столбца LOAN_NUM_CLOSED
merged_df['LOAN_NUM_CLOSED'] = merged_df.groupby('ID_CLIENT')['CLOSED_FL'].transform(lambda x: (x == 1).sum())

#Теперь можно удалить разные кредиты одного клиента
merged_df.drop_duplicates(subset=['ID_CLIENT'], keep='first', inplace=True)

#Отбираем нужные столбцы
columns = ['AGREEMENT_RK','TARGET','AGE','SOCSTATUS_WORK_FL','SOCSTATUS_PENS_FL','GENDER','CHILD_TOTAL','DEPENDANTS','PERSONAL_INCOME','LOAN_NUM_TOTAL','LOAN_NUM_CLOSED']
merged_df = merged_df.filter(columns)

In [4]:
merged_df.isna().sum()

AGREEMENT_RK         777
TARGET               777
AGE                    0
SOCSTATUS_WORK_FL      0
SOCSTATUS_PENS_FL      0
GENDER                 0
CHILD_TOTAL            0
DEPENDANTS             0
PERSONAL_INCOME      777
LOAN_NUM_TOTAL         0
LOAN_NUM_CLOSED        0
dtype: int64

In [5]:
merged_df["LOAN_NUM_TOTAL"].value_counts()

1     11288
2      2633
3       871
0       777
4       275
5       102
6        36
7        14
8         3
11        1
Name: LOAN_NUM_TOTAL, dtype: int64

#### Получается, что эти 777 строк с пропущенными значениями приходятся на клиентов, которые не брали кредитов

In [6]:
#Удаление строк с пропущенными значениями
merged_df = merged_df.dropna()

#Сбросим индексы
merged_df.reset_index(drop=True, inplace=True)

#Посмотрим на данные
merged_df

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
0        59910150.0     0.0   49                  1                  0   
1        59910230.0     0.0   32                  1                  0   
2        59910525.0     0.0   52                  1                  0   
3        59910803.0     0.0   39                  1                  0   
4        59911781.0     0.0   30                  1                  0   
...             ...     ...  ...                ...                ...   
15218    75291424.0     0.0   25                  1                  0   
15219    75291888.0     0.0   51                  1                  0   
15220    75291923.0     0.0   22                  1                  0   
15221    75291960.0     0.0   60                  0                  1   
15222    75292242.0     0.0   37                  1                  0   

       GENDER  CHILD_TOTAL  DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  \
0           1            2           1           5000.0               1   
1           1            3           3          12000.0               1   
2           1            4           0           9000.0               2   
3           1            1           1          25000.0               1   
4           0            0           0          12000.0               2   
...       ...          ...         ...              ...             ...   
15218       0            1           1          12000.0               3   
15219       1            2           0          10000.0               1   
15220       1            1           1           6000.0               1   
15221       1            2           0           4500.0               1   
15222       0            2           0          20000.0               1   

       LOAN_NUM_CLOSED  
0                    1  
1                    1  
2                    1  
3                    1  
4                    1  
...                ...  
15218                2  
15219                0  
15220                0  
15221                0  
15222                0  

[15223 rows x 11 columns]

In [8]:
# Сохранение DataFrame merged_df в формате CSV
merged_df.to_csv('merged_df.csv', index=False)

## Задание 2

При помощи инструмента Streamlit проведите разведочный анализ данных. В него может входить:

* построение графиков распределений признаков
* построение матрицы корреляций
* построение графиков зависимостей целевой переменной и признаков
* вычисление числовых характеристик распределения числовых столбцов (среднее, min, max, медиана и так далее)
* любые другие ваши идеи приветствуются!

[Пример Streamlit-приложения](https://rateyourflight.streamlit.app) с разведочным анализом, прогнозом модели и оценкой ее результатов.